### PR TITLE
Added RouteSort configurable class and Implementation.

### PR DESCRIPTION
**Pull request by merhatsidk**

### CODE REVIEW

File: onebusaway-watchdog-webapp/src/main/resources/org/onebusaway/watchdog/application-context-webapp.xml [hunk 1]
```diff
@@ -92,7 +92,7 @@
     </bean>
 	
 	<bean id="configurationServiceClient" class="org.onebusaway.util.impl.configuration.ConfigurationServiceClientFileImpl">
-		<constructor-arg type="java.lang.String" value="/var/lib/oba/config.json"/>
+		<constructor-arg type="java.lang.String" value="/opt/nyserda/tds/config.json"/>
```
Comment (sheldonabrown) [inline onebusaway-watchdog-webapp/src/main/resources/org/onebusaway/watchdog/application-context-webapp.xml]: Please review your local changes before committing, to prevent issues like this.  The default should stay /var/lib/oba/config.json

File: onebusaway-transit-data/src/main/java/org/onebusaway/transit_data/model/ArrivalsAndDeparturesQueryBean.java [hunk 2]
```diff
@@ -49,6 +49,8 @@ public final class ArrivalsAndDeparturesQueryBean implements Serializable {
   private FilterChain systemFilterChain = new FilterChain();
   private FilterChain instanceFilterChain = new FilterChain();
 
+  private RouteSort subwayRouteSort = new RouteSort("1,2,3,4,5,6,7,A,C,E,B,D,F,M,G,L,N,Q,R,W,J,Z,Sf,S,Sr,SIR");
```
Comment (sheldonabrown) [inline onebusaway-transit-data/src/main/java/org/onebusaway/transit_data/model/ArrivalsAndDeparturesQueryBean.java]: I'm not comfortable with this code here.  Can you remove it to configuration, the default will instead be an alphanumeric sort?

File: onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/search/RouteAction.java [hunk 3]
```diff
@@ -44,7 +49,13 @@ public DefaultHttpHeaders index() throws IOException, ServiceException {
       BeanFactoryV2 factory = getBeanFactoryV2();
       RouteSearchResultBean result = new RouteSearchResultBean();
       result.setSuggestions(routeSuggestions);
-      return setOkResponse(factory.getResponse(result));
+      ListWithReferencesBean<RouteV2Bean> factoryResponse = factory.getResponse(result);
+//      factoryResponse.
```
Comment (sheldonabrown) [inline onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/search/RouteAction.java]: please remove dead code

File: onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/ArrivalsAndDeparturesForLocationAction.java [hunk 4]
```diff
@@ -58,6 +60,8 @@ public void setFilterChain(FilterChain filterChain) {
         _query.setSystemFilterChain(filterChain);
     }
 
+    @Autowired
+    private RouteSort subwayRouteSort;
```
Comment (sheldonabrown) [inline onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/ArrivalsAndDeparturesForLocationAction.java]: lets not call this subwayRouteSort.  We have a mental model issue here.

If RouteSort is always going to be present, then it should be defaultSort.  If RouteSort is only optionally present, then we need to test for null and not sort if empty

File: onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/ArrivalsAndDeparturesForLocationAction.java [hunk 5]
```diff
@@ -164,7 +168,17 @@ public DefaultHttpHeaders index() throws IOException, ServiceException {
         if (adResult == null) {
             return emptyResponse();
         }
-        return setOkResponse(factory.getResponse(adResult));
+
+        EntryWithReferencesBean<StopsWithArrivalsAndDeparturesV2Bean> response = factory.getResponse(adResult);
+
+        response.getEntry().getArrivalsAndDepartures()
```
Comment (sheldonabrown) [inline onebusaway-api-webapp/src/main/java/org/onebusaway/api/actions/api/where/ArrivalsAndDeparturesForLocationAction.java]: I get that this is super convenient to do right here, but it feels wrong.  If you moved it into the BeanFactory, couldn't it be re-used?  I could be wrong -- just asking.  

Basically because you decided to do this here, you need to copy this code fragment to N other places, violating the DRY prinicipal

**Review by sheldonabrown — CHANGES_REQUESTED** [review]
A good first pass!  Please look at my suggestions to see if things can be improved.

File: onebusaway-api-core/src/main/java/org/onebusaway/api/model/transit/BeanFactoryV2.java [hunk 6]
```diff
@@ -159,12 +195,49 @@ public EntryWithReferencesBean<BlockV2Bean> getBlockResponse(BlockBean block) {
...
           StopsWithArrivalsAndDeparturesBean result) {
-      return entry(getStopsWithArrivalAndDepartures(result));
+
+    EntryWithReferencesBean<StopsWithArrivalsAndDeparturesV2Bean> response = entry(getStopsWithArrivalAndDepartures(result));
+
+    String agencyId = response
+            .getEntry()
+            .getArrivalsAndDepartures().size() > 0 ?
+            response.getEntry().getArrivalsAndDepartures().get(0)
+                    .getRouteId().split("_")[0] :
```
Comment (sheldonabrown) [inline onebusaway-api-core/src/main/java/org/onebusaway/api/model/transit/BeanFactoryV2.java]: Try not to assume the agencyAndId separator will be "_".  Use AgencyAndIdLibrary to convert

File: onebusaway-api-webapp/src/main/resources/data-sources.xml [hunk 7]
```diff
@@ -86,4 +86,12 @@
             </props>
         </property>
     </bean>
+    <bean id="customRouteSort" class="org.onebusaway.transit_data.model.RouteSort">
```
Comment (sheldonabrown) [inline onebusaway-api-webapp/src/main/resources/data-sources.xml:89]: Question:  does customRouteSort always need to be defined?  Should there be an emptyRouteSort that's an no-op?

File: onebusaway-api-webapp/src/main/resources/data-sources.xml [hunk 8]
```diff
@@ -86,4 +86,12 @@
...
+        <constructor-arg name="agencySortConfiguration" >
+            <map>
+                <entry key="MTASBWY" value="1,2,3,4,5,6,7,A,C,E,B,D,F,M,G,L,N,Q,R,W,J,Z,Sf,S,Sr,SIR"/>
```
Comment (sheldonabrown) [inline onebusaway-api-webapp/src/main/resources/data-sources.xml]: Let's not commit this here, as it will apply to all deployments.  This fragment belongs in dev-ops, I'll show you where

**Review by sheldonabrown — CHANGES_REQUESTED** [review]
It looks really good now.  Just two small things to clean up, and a question as to whether a no-op default sort bean should exist to short-cut the usual procedure.

File: onebusaway-api-core/src/main/java/org/onebusaway/api/model/transit/BeanFactoryV2.java (hunk 6 above)
Comment (merhatsidk) [reply onebusaway-api-core/src/main/java/org/onebusaway/api/model/transit/BeanFactoryV2.java]: Sure, I will fix that.

File: onebusaway-api-webapp/src/main/resources/data-sources.xml (hunk 7 above)
Comment (merhatsidk) [reply onebusaway-api-webapp/src/main/resources/data-sources.xml:89]: The compareRoutes method in RouteSort class will check whether configuration is provided and if not it will use  a standard sort so no need of an emptyRouteSort with no-op.

File: onebusaway-api-webapp/src/main/resources/data-sources.xml (hunk 8 above)
Comment (merhatsidk) [reply onebusaway-api-webapp/src/main/resources/data-sources.xml]: Sure.
